cms/server/ContestWebServer.py

```diff
@@ -1061,10 +1063,12 @@ def which_language(user_filename):
                              if it is not a recognized language.
 
             """
-            for source_ext, language in SOURCE_EXT_TO_LANGUAGE_MAP.iteritems():
-                if user_filename.endswith(source_ext):
-                    return language
-            return None
+
+            languageSubmitted = self.get_argument("submission_lang", "")
```

Comment:
The default of "" is wrong: either we should be making a best-guess at the submission language if "submission_lang" is not set (so use the previous code), or you should just be returning None since the calling code to which_language expects None to mean "whoops user is a moron".